### PR TITLE
fix(components): [tree] dragging a node will deselect the node

### LINKED ISSUE #13912: [Component] [tree] el-tree，同时开启多选框和拖拽功能后，拖拽节点会导致节点取消选中
Bug Type: **`Component`**

## Environment

- Vue Version: `3.3.4`
- Element Plus Version: `2.3.9`
- Browser / OS: `Microsoft Edge 版本 115.0.1901.188 (正式版本) (64 位)`
- Build Tool: `Vite`

### Related Component

- `el-tree`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6IjxzY3JpcHQgc2V0dXAgbGFuZz1cInRzXCI+XG5jb25zdCBkYXRhID0gW1xuICB7IGlkOiAxLCBsYWJlbDogJ2FhYScgfSxcbiAgeyBpZDogMiwgbGFiZWw6ICdiYmInIH0sXG4gIHsgaWQ6IDMsIGxhYmVsOiAnY2NjJyB9LFxuXVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGVsLXRyZWVcbiAgICAgIG5vZGUta2V5PVwiaWRcIlxuICAgICAgOmRhdGE9XCJkYXRhXCJcbiAgICAgIDpkZWZhdWx0LWNoZWNrZWQta2V5cz1cIlsxLDIsM11cIlxuICAgICAgOnByb3BzPVwie1xuICAgICAgICBjaGlsZHJlbjogJ2NoaWxkcmVuJyxcbiAgICAgICAgbGFiZWw6ICdsYWJlbCcsXG4gICAgICB9XCJcbiAgICAgIGRyYWdnYWJsZVxuICAgICAgc2hvdy1jaGVja2JveFxuICAgIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWUsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnt9fQ==)

### Steps to reproduce

el-tree，同时开启多选框和拖拽功能后，拖拽节点会导致节点取消选中

### What is Expected?

拖拽节点后保持原本的checked状态

### What is actually happening?

节点选中状态被取消

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by Alixhan**
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

## Description

同时开启多选框和拖拽功能后，拖拽节点会导致节点取消选中

## Related Issue

Fixes #13912.

## Explanation of Changes

拖拽节点时插入新位置时，将checked状态同步带过去即可

### CODE REVIEW

**Review by btea — APPROVED** [review]
LGTM

**Comment by boomboy4** [comment]
@btea I've tried this before, but it causes the parent and child nodes before and after the move to have incorrect checkboxes.
I don't think that's right.

**Comment by boomboy4** [comment]
Like this

**Comment by Alixhan** [comment]
@boomboy4 I just found another bug, unrelated to this PR, which causes the checkbox state of the parent node of the changing node to not be updated, Let me fix it.

**Comment by Alixhan** [comment]
Are you talking about that?

**Comment by boomboy4** [comment]
> Are you talking about that?
I mean you only change the state of the dragging node, but you don't manage the old parent-child and new parent-child states of the dragging node

**Comment by boomboy4** [comment]
> Are you talking about that?

I mean you only change the state of the dragging node, but you don't manage the old parent-child and new parent-child states of the dragging node

**Comment by boomboy4** [comment]
a6a49aed7782e4ee3d7d590a1c8db04
Before drag

**Comment by boomboy4** [comment]
1ade5ce04a2df614b07226c823af724
dragged

**Comment by Alixhan** [comment]
@boomboy4 I know your mean，but it is another bug，It existed before I submitted this PR and I just found out about it.

**Comment by Alixhan** [comment]
@boomboy4 I'm trying to fix it

**Comment by boomboy4** [comment]
> @boomboy4 I'm trying to fix it

ok. the pr supposed to solve the problem of state after dragging and dropping

**Comment by Alixhan** [comment]
@boomboy4 @btea I added a listener to  `tree-node` component to watch for changes to the childNode's length，It can also solve the problem of [#14915](https://github.com/element-plus/element-plus/pull/14915)

**Comment by boomboy4** [comment]
> @boomboy4 @btea I added a listener to `tree-node` component to watch for changes to the childNode's length，It can also solve the problem of [#14915](https://github.com/element-plus/element-plus/pull/14915)

This doesn't solve the problem because the listen here only changes the current node, the state above the parent doesn't change.
I don't have permission to review the code though, that's @btea  job. But thanks for your efforts!

**Comment by Alixhan** [comment]
@boomboy4 What you may not have noticed is that changing the checked state triggers the checked listener event, which automatically triggers an update to the parent state, I've tested it and it works fine.

**Comment by Alixhan** [comment]
Before
image
dragged
image

**Comment by boomboy4** [comment]
> @boomboy4 What you may not have noticed is that changing the checked state triggers the checked listener event, which automatically triggers an update to the parent state, I've tested it and it works fine.

But darggingNode parent's parent, which has no change in the number of children, will not respond.
You can try dragging to deeper levels, the top level won't change state

**Comment by Alixhan** [comment]
@boomboy4 Yes, I noticed. This line("tree-node.vue:332")  solves that problem.

**Comment by boomboy4** [comment]
> @boomboy4 Yes, I noticed. This line("tree-node.vue:332") solves that problem.

Yes, but setChecked is limited to checked or unchecked, and node.indeterminate changes do not affect the parent's indeterminate.

**Comment by boomboy4** [comment]
> @boomboy4 Yes, I noticed. This line("tree-node.vue:332") solves that problem.

77daa60a5d2ba1801b3450ba6388bf1

b00c7cf76627b68da7cb3d2edd6ce60

**Comment by Alixhan** [comment]
@boomboy4 Thank you for recognizing that I have come to understand the importance of being 100% meticulous in resolving issues. HA HA HA^^^

**Comment by Nullaha** [comment]
> @boomboy4 Thank you for recognizing that I have come to understand the importance of being 100% meticulous in resolving issues. HA HA HA^^^

我看了一下你的提交，确实解决了https://github.com/element-plus/element-plus/pull/14915

**Comment by Alixhan** [comment]
@yujinpan @ryuhangyeong 为啥不给审批呢？

**Comment by btea** [comment]
@Alixhan Hello, the project is driven by the community, and team members all have their main work to do. Recently, it may be that due to busy work, there is less time to maintain the project. 
Therefore, some PRs may be missed or may not receive a response for a long time. Please wait patiently. ❤️

**Comment by btea** [comment]
@Alixhan Could you add a test case?

**Comment by Alixhan** [comment]
> @Alixhan Could you add a test case?

节点的拖动无法模拟，并且这个是bug修改，前面功能就没有测试用例，有点无从下手。

**Comment by btea** [comment]
In the scenario provided by #15477, dragging a node seems to cause an error.

**Comment by Alixhan** [comment]
> In the scenario provided by #15477, dragging a node seems to cause an error.
此PR正是解决了这个问题。

**Comment by btea** [comment]
For the scenario provided by #15477, the content you modify in this PR will cause an error.

play

**Comment by Alixhan** [comment]
> For the scenario provided by #15477, the content you modify in this PR will cause an error.

我看了一下，这个tree组件设计之初就是在拖动后，节点对象会被浅复制，并且丢失选中状态，节点的id也会重新生成，要保持拖动后节点的选中状态，必须设置`node-key`属性，否则无法找到拖动后复制的节点设置选中状态.

**Comment by Alixhan** [comment]
@btea 抽空合并一下这个PR呗，着急用。

**Review by kooriookami — APPROVED** [review]

**Comment by btea** [comment]
> @btea 抽空合并一下这个PR呗，着急用。

合并发版可能没那么快，急用的话，建议先自行patch。

**Comment by kooriookami** [comment]
> @kooriookami Do you think there are any other problems with this? If not, I think it can be merged.

To be honest, I'm not sure if it'll cause any other problems :D

File: packages/components/tree/src/model/useDragNode.ts [hunk 1]
```diff
@@ -180,6 +180,13 @@ export function useDragNodeHandler({ props, ctx, el$, dropIndicator$, store }) {
       }
       if (dropType !== 'none') {
         store.value.registerNode(draggingNodeCopy)
+        //restore checkbox state after dragging
+        draggingNode.node.eachNode((node) => {
+          store.value.nodesMap[node.data[store.value.key]]?.setChecked(
```
Comment (btea) [inline packages/components/tree/src/model/useDragNode.ts:185]: Maybe we should be compatible with the case where `store.node.key` is `undefined`.

Comment (Alixhan) [reply packages/components/tree/src/model/useDragNode.ts:185]: 可以这样，当store.node.key为undefined，不执行多选框的恢复操作。

Comment (Alixhan) [reply packages/components/tree/src/model/useDragNode.ts:185]: 我来修改一下

**Review by btea — APPROVED** [review]
@kooriookami  There is no problem so far, we can merge it first.

**Pull request merged**